components/MaskedTransition/src/MDCMaskedTransition.m

```diff
@@ -0,0 +1,269 @@
+  return CGPointMake(CGRectGetMidX(frame), CGRectGetMidY(frame));
+}
+
+static inline CGPoint AnchorPointFromPosition(CGPoint position, CGRect bounds) {
+  return CGPointMake(position.x / bounds.size.width, position.y / bounds.size.height);
+}
+
+static inline CGRect FrameCenteredAround(CGPoint position, CGSize size) {
+  return CGRectMake(position.x - size.width / 2,
+                    position.y - size.height / 2,
+                    size.width,
+                    size.height);
+}
+
+static inline CGFloat LengthOfVector(CGVector vector) {
+  return (CGFloat)sqrt(vector.dx * vector.dx + vector.dy * vector.dy);
+}
+
+@interface MDCMaskedTransition () <MDMTransitionWithPresentation, MDMTransitionWithFallback>
+@end
+
+@implementation MDCMaskedTransition {
+  UIView *_sourceView;
+  MDCMaskedPresentationController *_presentationController;
+  BOOL _shouldSlideWhenCollapsed;
+}
+
+- (instancetype)initWithSourceView:(UIView *)sourceView {
+  self = [super init];
+  if (self) {
+    _sourceView = sourceView;
+  }
+  return self;
+}
+
+- (id<MDMTransition>)fallbackTransitionWithContext:(id<MDMTransitionContext>)context {
+  return _shouldSlideWhenCollapsed ? nil : self;
+}
+
+#pragma mark - MDMTransitionWithPresentation
+
+- (UIModalPresentationStyle)defaultModalPresentationStyle {
+  return UIModalPresentationCustom;
+}
+
+- (UIPresentationController *)presentationControllerForPresentedViewController:(UIViewController *)presented
+                                                      presentingViewController:(UIViewController *)presenting
+                                                          sourceViewController:(UIViewController *)source {
+  _presentationController =
+      [[MDCMaskedPresentationController alloc] initWithPresentedViewController:presented
+                                                      presentingViewController:presenting
+                                                 calculateFrameOfPresentedView:_calculateFrameOfPresentedView];
+  return _presentationController;
+}
+
+- (void)startWithContext:(NSObject<MDMTransitionContext> *)context {
+  MDCMaskedTransitionMotionSpec spec = motionForContext(context);
+  if (context.direction == MDMTransitionDirectionForward) {
+    _shouldSlideWhenCollapsed = spec.shouldSlideWhenCollapsed;
+  }
+
+  MDMMotionAnimator *animator = [[MDMMotionAnimator alloc] init];
+  animator.shouldReverseValues = context.direction == MDMTransitionDirectionBackward;
+
+  // Cache original state.
+  // We're going to reparent the fore view, so keep this information for later.
+  UIView *originalSuperview = context.foreViewController.view.superview;
+  const CGRect originalFrame = context.foreViewController.view.frame;
+  UIView *originalSourceSuperview = _sourceView.superview;
+  const CGRect originalSourceFrame = _sourceView.frame;
+  UIColor *originalSourceBackgroundColor = _sourceView.backgroundColor;
+
+  UIView *scrimView;
+  if (!_presentationController) {
```

Comment:
The transitioning delegate is queried for an animation and interaction controller for any type of modal presentation style. 

> When you want to present a view controller using a custom modal presentation type, set its modalPresentationStyle property to custom and assign an object that conforms to this protocol to its transitioningDelegate property. When you present that view controller, UIKit queries your transitioning delegate for the objects to use when animating the view controller into position.

https://developer.apple.com/documentation/uikit/uiviewcontrollertransitioningdelegate